object/object_test.go

```diff
@@ -25,14 +25,166 @@ func TestTombstone(t *testing.T) {
+	require.False(t, v1.Stomps(nil), "should treat nil as the zero version")
+	require.False(t, v1.Stomps(&Version{Pid: 8, Version: 1}), "version should not stomp equal version")
+
+	// Linear from v1 but with different PIDs (v2 stomps v3)
+	v2 := &Version{Pid: 8, Version: 2, Parent: &Version{Pid: 8, Version: 1}}
+	v3 := &Version{Pid: 9, Version: 2, Parent: &Version{Pid: 8, Version: 1}}
+
+	require.True(t, v2.Stomps(v3), "stomps should be detected for concurrent versions")
+	require.False(t, v3.Stomps(v2), "only one concurrent version should stomp the other")
+	require.False(t, v2.Stomps(v2), "version should not stomp itself")
+
+	// Linear from v3 (should not stomp v2 even though it's later)
+	v4 := &Version{Pid: 9, Version: 3, Parent: &Version{Pid: 9, Version: 2}}
+	require.False(t, v4.Stomps(v2), "later version should not stomp earlier version if not concurent")
+	require.False(t, v4.Stomps(v3), "version should not stomp its parent")
+}
+
+func TestSkips(t *testing.T) {
+	// Root version
+	v1 := &Version{Pid: 8, Version: 1, Parent: nil}
+
+	require.False(t, v1.Skips(&VersionZero), "root version should not skip zero value")
+	require.False(t, v1.Skips(nil), "should treat nil as the zero version")
+	require.False(t, v1.Skips(&Version{Pid: 8, Version: 1}), "version should not skip equal version")
+
+	// Linear from v1 but with different PIDs (v2 stomps v3)
+	v2 := &Version{Pid: 8, Version: 2, Parent: &Version{Pid: 8, Version: 1}}
+	v3 := &Version{Pid: 9, Version: 2, Parent: &Version{Pid: 8, Version: 1}}
+
+	require.False(t, v2.Skips(v1), "version should not skip its parent")
+	require.False(t, v2.Skips(v3), "stomp should not be a skip")
+	require.False(t, v3.Skips(v2), "concurrent version should not be a skip")
+
+	// Linear from v3 (skips v3 from v1)
+	v4 := &Version{Pid: 9, Version: 3, Parent: &Version{Pid: 9, Version: 2}}
+	require.True(t, v4.Skips(v1), "a later version should skip an earlier version if parent is not equal")
+	require.True(t, v4.Skips(v2), "a branch should be considered a skip if the parent was stomped")
+	require.False(t, v4.Skips(v3), "a child should not skip a non-root parent")
+}
+
+func TestVersionHistory(t *testing.T) {
+	// Create a version history without reusing pointers
+	v71 := &Version{Pid: 7, Version: 1, Parent: nil}
+	v72 := &Version{Pid: 7, Version: 2, Parent: &Version{Pid: 7, Version: 1}}
+	v73 := &Version{Pid: 7, Version: 3, Parent: &Version{Pid: 7, Version: 2}}
+	v83 := &Version{Pid: 8, Version: 3, Parent: &Version{Pid: 7, Version: 2}}
+	v74 := &Version{Pid: 7, Version: 4, Parent: &Version{Pid: 7, Version: 3}}
+	v84 := &Version{Pid: 8, Version: 4, Parent: &Version{Pid: 8, Version: 3}}
+	v85 := &Version{Pid: 8, Version: 5, Parent: &Version{Pid: 8, Version: 4}}
+	v95 := &Version{Pid: 9, Version: 5, Parent: &Version{Pid: 7, Version: 4}}
+	v76 := &Version{Pid: 7, Version: 6, Parent: &Version{Pid: 9, Version: 5}}
+	v97 := &Version{Pid: 9, Version: 7, Parent: &Version{Pid: 7, Version: 6}}
```

Comment:
A rough sketch of this version history:

![Paper Journal 14](https://user-images.githubusercontent.com/745966/151077837-56506ab9-a623-41df-8b2e-69c55daef282.png)